pip/pip-269.md

```diff
@@ -0,0 +1,239 @@
+| 2 | | Waiting for the response of BK |
+| 3 | Clear redelivery messages |
+| 4 | Set read position to `3:0` |
+| 5 | | Read complete, set the read position to `3:7`, and send `[3:5 ~ 3:9]` to the client |
+
+In this scenario(referred to as **Scenario-2**), the consumer will receive `[3:5 ~ 3:9]` first and the messages `[3:0~3:4]` will not be delivered to the client until unloading the topic.
+
+---
+
+In **Occasional Read Position Change** is executed, the attributes(including `read position`) of the Cursor will be changed(`action 1`)<sup>[2]</sup>, and the attributes of the Dispatcher(`action 2`) will be changed too<sup>[3]</sup>. If these two actions are not executed atomicity, this will cause the attributes of Cursorto to be inconsistent with the attributes of Disptacher, and cause some problems. For example:
+| time | **Reset cursor** | **Consumer reconnect** |
+| --- | --- | --- |
+| 1 | disconnect all consumers |
+| 2 | | remove consumer |
+| 3 | | add consumer_1 |
+| 4 | | rewind cursor to `3:0` |
+| 5 | | add consumer_2 |
+| 6 | | add consumer_2 to recently joined consumers, `max read position` is `3:0`<sup>[4]</sup> |
+| 7 | set read position to `1:0` |
+
+In this scenario(referred to as **Scenario-3**), the `consumer_2` will be stuck by the mechanism Recently Joined Consumers of Key_Shared mode<sup>[4]</sup>. It's a simple question, but it's more complicated than that:
+- If there has an in-flight reading.
+- If there is more than one **Occasional Read Position Change** action running.
+- Both above.
+
+---
+
+In the Multi Consumer Dispatcher, there has a mechanism that tries to fix the issue caused by the action `cursor.rewind` and **Sequentially Read** executed concurrently(referred to as Rewind After In-flight Reading). It works like this: if there has an in-flight reading when calling `cursor.rewind`, just do rewind after the in-flight reading is finished.` but this mechanism has the following drawbacks:
+- Makes a new issue lead to a new issue<sup>[5]</sup>.
+- Only applicable to `cursor.rewind`.
+- Only applicable to Multi Consumer Dispatcher.
+- If there is more than one in-flight reading(one Sequentially Read and some Message Replay Read might exist simultaneously), this mechanism can not work.
+
+# Goals
+- For the **Scenario-1** and **Scenario-2**(see Motivation), discard the in-flight reading(referred to as **Goal-1**).
+- For the **Scenario-3**(see Motivation), we should do these two things:
+    - Provide a way to guarantee the atomicity modification of both attributes of Cursor and attributes of Dispatcher(referred to as **Goal-2**).
+    - Prevent more than one **Occasional Read Position Change** executed at the same time(Current Proposal does not consider this optimization, there will be some bug fixes after this PIP)
+- Revert https://github.com/apache/pulsar/issues/6403
+- Remove the mechanism Rewind After In-flight Reading<sup>[5]</sup>
+
+# High-Level Design
+
+**For the Goal-1**: Cursor Epoch
```

Comment:
Ok. First, let me try to phrase your idea in my words. Tell me if I'm correct:

You're basically adding a variable to a cursor, named (my name, of course) `readPositionExternalResetCount`. This variable will be increase when ever the read position is about to be reset to some value, initiated by an external action like reset-subscription, rewind, etc.

Every time a read sequence begins, it starts by recording the readPositionExternalResetCount. When the reading has finished and read position is about to be updated, it verifies no external reset has been done by verifying the `readPositionExternalResetCount` was not increased relative to the recorded value. If it has, this means a reset action has started, which means we can drop and stop the reading sequence we have started. If not, no reset was done, continue and update read position. 
Please note even this is not transactional and fully safe. 

What do I suggest:
1. Rename variable. Epoch in computer science is mostly for how many seconds has passed since 1970-01-01. See https://en.wikipedia.org/wiki/Epoch_(computing)
2. Checking count has increased and if not update read position is not fully safe. Better guard update read position with a lock, since you're taking an action based on the count. I would take a lock on read position, and in that lock update count if needed and read position if needed.
3. I wouldn't use an exception for flow control. It's an expensive operation, especially for latency sensitive systems. Return value like Result of some sort - can we do that?

Reply:
> Tell me if I'm correct

Correct.

> Rename variable. Epoch in computer science is mostly for how many seconds have passed since 1970-01-01. See https://en.wikipedia.org/wiki/Epoch_(computing)

Pulsar and Zookeeper have similar concepts named epoch, so `epoch` may be easier to understand? I have also corrected the document: `epoch` is also related to the change of `mark deleted position`(in other words, if the `mark deleted position` is changed by Cursor Reset, the `epoch` also increases)

> Checking count has increased and if not update read position is not fully safe. Better guard update read position with a lock, since you're taking an action based on the count. I would take a lock on read position, and in that lock update count if needed and read position if needed.

Yes, this is consistent with the current design, the original document may not be clearly enough, I improved it

> I wouldn't use an exception for flow control. It's an expensive operation, especially for latency sensitive systems. Return value like Result of some sort - can we do that?

Good suggestion, already instead of `return -1`.

Reply:
> Pulsar and Zookeeper have similar concepts named epoch, so epoch may be easier to understand? I have also corrected the document: epoch is also related to the change of mark deleted position(in other words, if the mark deleted position is changed by Cursor Reset, the epoch also increases)

ZK epoch seems more like a revision or version or sequence ID. 
Given your explanation, it seems you need a revision/version/sequence ID for the state of the cursor state. How about `cursorStateRevisionNum`?

Revision is:

> The act of changing or correcting something, or the thing that has been changed or corrected: [ U ] The Senate is expected to act on tax revision.

Reply:
Good suggestion. already change this variable to `cursorStateRevisionNum `